CARDS-2580 - Number question: allow min/max value to be specified but not enforced

marta- opened the pull request:
Includes a test questionnaire `Number Min / Max Value Test`.

marta- committed:
CARDS-2580 - Number question: allow min/max value to be specified but not enforced

marta- committed:
CARDS-2580 - Number question: allow min/max value to be specified but not enforced

Enable marckdown in validation messages

sashaandjic reviewed (approved):
tested

![image](https://github.com/user-attachments/assets/8b492acd-d6c6-40d5-af8f-c20b7e5db8ae)

Upon save, the form is not marked as `Draft` and `Invalid` if min/max are not enforced.

marta- committed:
CARDS-2580 - Number question: allow min/max value to be specified but not enforced

Addressing review comments - use the default message when min/max values are enforced.

sashaandjic commented:
When enforcing min/max is enabled, the error msg shows up:
- in black fonts if withing the range
- in red fonts of not within range

![image](https://github.com/user-attachments/assets/8baed143-4016-459f-9f70-815f4e954cf3)

![image](https://github.com/user-attachments/assets/19a07d0d-2c71-4028-93d3-56979ce3aaa1)

This is due to the existing behavior, when error msg is blank and we display the old message,

sashaandjic reviewed (approved):
tested

marta- commented:
> When enforcing min/max is enabled, the error msg shows up:
> 
> * in black fonts if withing the range
> * in red fonts of not within range

The code has been updated in 2bb40f1 to have the exact same behavior as before (no custom message).

sashaandjic commented:
![image](https://github.com/user-attachments/assets/3cb33016-7c5a-4950-8bfc-44e103d3c9e6)

It would be nice to always display range message below the input line.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -133,6 +133,8 @@
             "maxAnswers": "long",
             "minValue": "double",
             "maxValue": "double",
```
Why are the min-max values for a `long` question `double`?

marta- replied:
Because it's the same property regardless if the dataType is long, double, or decimal. We can't specify different types for different values of dataType.

sdumitriu reviewed (commented):
This behaves slightly different from enforced ranges:
- In enforced ranges, the text is validated as it is typed, so if minValue=10, typing 2 will trigger the error, typing 20 will remove the error. In optional ranges, the error is triggered only when the input loses focus, but the error is removed while typing if the value becomes valid.
- In optional ranges, the bottom line of the input also turns red, which I like, we should do that for enforced ranges as well.
- If the error message is displayed, deleting the whole value doesn't clear the error message, it still appears to be invalid, but in enforced ranges an empty input is considered valid (if minAnswers=0).
- Is this a regression, or was it always like this? If I "Hide answer instructions", the error message for enforced ranges is never shown.

marta- committed:
CARDS-2580 - Number question: allow min/max value to be specified but not enforced

Addressing review comments - clear the error for soft limits when the value is empty

marta- committed:
CARDS-2580 - Number question: allow min/max value to be specified but not enforced

Addressing review comments - give the input an error state (red border) when the value is outside enforced limits

marta- commented:
> This behaves slightly different from enforced ranges:
> 
> * In enforced ranges, the text is validated as it is typed, so if minValue=10, typing 2 will trigger the error, typing 20 will remove the error. In optional ranges, the error is triggered only when the input loses focus, but the error is removed while typing if the value becomes valid.

I can add the `liveValidation` property and make the behavior configurable. Given the "soft" nature of this validation, checking as you type didn't make sense at the time.

> * In optional ranges, the bottom line of the input also turns red, which I like, we should do that for enforced ranges as well.

Updated.

> * If the error message is displayed, deleting the whole value doesn't clear the error message, it still appears to be invalid, but in enforced ranges an empty input is considered valid (if minAnswers=0).

Updated.

> * Is this a regression, or was it always like this? If I "Hide answer instructions", the error message for enforced ranges is never shown.

It has always been like this.

sashaandjic reviewed (approved):
tested - looks good

sdumitriu reviewed (commented):
Still a minor difference: when opening a new form with no answers, the unenforced answers start as invalid. Entering a value and deleting it makes it valid again.

marta- committed:
CARDS-2580 - Number question: allow min/max value to be specified but not enforced

Addressing review comments: fixed "when opening a new form with no answers, the unenforced answers start as invalid"

sdumitriu reviewed (changes requested):
Not as expected when multiple values are allowed: Long question, maxAnswers=0, range for the values. When the range is enforced, pressing Enter still adds an invalid value as one of the answer values, but when the range is not enforced, Enter doesn't accept the value, but Tab or click outside the input will still select it. This is a regression for enforced questions, and it should be the other way around.

If the minValue is >=0, then the input field doesn't allow entering negative values. I think we should disable the "no-negatives" behavior when the range is not enforced. 

Also with multiple answers, each invalid answer gets the out-of-range message appended. I think it's a good choice, just raising it in case it's not what you want.
![image](https://github.com/user-attachments/assets/5c4d685a-db63-4d9a-9a26-1296edd1eef3)

marta- committed:
CARDS-2580 - Number question: allow min/max value to be specified but not enforced

Addressing review comments:
* Bugfix: soft min/max values are enforced for multivalued answers
* Regression: hard min/max values are no longer enforced for multivalued answers

marta- committed:
CARDS-2580 - Number question: allow min/max value to be specified but not enforced

Addressing review comments: allow negatives regardless of minValue if not enforced

marta- commented:
> Not as expected when multiple values are allowed: Long question, maxAnswers=0, range for the values. When the range is enforced, pressing Enter still adds an invalid value as one of the answer values, but when the range is not enforced, Enter doesn't accept the value, but Tab or click outside the input will still select it. This is a regression for enforced questions, and it should be the other way around.

Addressed in 2fea07e .

> If the minValue is >=0, then the input field doesn't allow entering negative values. I think we should disable the "no-negatives" behavior when the range is not enforced.

Adressed in 806b87f .

> Also with multiple answers, each invalid answer gets the out-of-range message appended. I think it's a good choice, just raising it in case it's not what you want. 

This is what we want.

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (approved):
This works well enough, on par with dev. Some things noticed while testing:
- number, single value, enforced range: entering an out-of-range number will save this invalid value when saving the form
- number, multiple values, enforced range: entering an out-of-range number and leaving the field does not accept the value, and does not save it when saving the form
- text, multiple values, validation regexp: entering a non-matching value will still accept and save it
- in view mode out-of-range numerical values aren't displayed as invalid, but non-matching text values do display the invalid-regexp message

For consistency, we should allow out-of-range numerical values to be saved if entered, and we should display the error/warning messages in view mode.

sashaandjic commented:
> For consistency, we should allow out-of-range numerical values to be saved if entered, and we should display the error/warning messages in view mode.

+1
